### ISSUE TITLE
VMware Fusion: Host volume file corruption

### ISSUE THREAD

**Issue by adambiggs** [docker/machine#1460]
I've been testing Machine with VMware Fusion for local development now that #641 has been resolved, but have run into another issue with file corruption in host volumes.

When writing to an existing file on the host, in some cases the shared file in the VM is corrupted. Similar to #1275, but instead of the file having no contents, it contains the *old* data (before being written), but it's length is updated to match that of the host file...

- If characters are removed from host file, shared file is truncated to match same amount of characters
- If characters are added to host file, `^@` null characters are appended to shared file
- In both cases, shared file contents are not updated with actual changes from host file

Strangely, I only noticed this when compiling CoffeeScript files. Manually writing to a file with Vim works totally fine. Not sure why it works with Vim... but even a simple `$ echo "test 123" > test.js` on the host will trigger the problem.

I was able to reliably reproduce the issue with these steps:
- Create `test.coffee` with some dummy content and compile it on host
    - `test.js` looks good in VM

- Add or remove some lines from `test.coffee` and recompile
    - `test.js` is now corrupt in VM

- Open `test.js` with host Vim and save
    - `test.js` is fixed in VM

- `$ echo "test 123" > test.js` on host
    - `test.js` is corrupt again in VM

Very strange...

**Comment by adambiggs** [comment]
I found this thread in the VMware forums which sounds very similar: https://communities.vmware.com/thread/462303

**Comment by donnex** [comment]
I'm having the same issue. My python files got corrupt when I was testing docker-machine with Fusion. I had a shared folder from my Yosemite host to my VMWare Fusion Docker VM. Noticed some really strange file corruptions/files not updating when saved.

I was running Gunicorn inside the container and it picked up and reloaded the server when files were saved, but the content of the files did not update.

The only way I found to get the files to update was to restart the VM.

**Comment by nathanleclaire** [comment]
cc @frapposelli

**Comment by frapposelli** [comment]
@adambiggs @donnex I tried but cannot reproduce, can you get me a simple case I can test? also, which version of Docker/Boot2Docker are you using?

**Comment by donnex** [comment]
## Versions
Docker version 1.7.0, build 0baf609

docker-machine version 0.3.0 (HEAD)

docker-compose version: 1.3.1
CPython version: 2.7.6
OpenSSL version: OpenSSL 0.9.8zf 19 Mar 2015

VMWare Fusion Version 7.1.2 (2779224)

Boot2Docker version 1.7.0, build master : 7960f90

## To reproduce
Here is how I can reproduce this error. Creating a standard new docker-machine and then running Nginx with a index.html file mounted from a shared folder.

mkdir fusion-test
cd fusion-test
docker-machine create -d vmwarefusion fusion-test
eval "$(docker-machine env fusion-test)"
echo "test" > index.html
docker run --rm -p 80:80 --name nginx-fusion-test -v $(pwd):/usr/share/nginx/html:ro nginx

*Check http://DOCKER_MACHINE_IP should say test*

echo "test2" > index.html

*Check http://DOCKER_MACHINE_IP should say test2, first update seems to work*

echo "test3" > index.html

*Check http://DOCKER_MACHINE_IP should say test3 but here mine says test2 still and the only way I've found to make it refresh the file is to restart the docker-machine*

**Comment by frapposelli** [comment]
@donnex thanks, I'm able to reproduce now, seems like `vmhgfs` is not picking up changes if the filesize stays the same (see https://asciinema.org/a/bq8qrusg4pclg8qeewrw17sqs).

I will file a bug for this as it needs to be fixed upstream.

**Comment by donnex** [comment]
@frapposelli Ah, sweet, let's hope it gets fixed.

Could that problem also explain the corrupted files? I was able to reproduce the corrupted files in a similar way as the previous one with files not updating.

If you start with a large index.html, I added ~1000 lines of lorem ipsum. Then add a word to search for at the end of this file, I used ENDING. As far as I can tell the file get corrupted if you remove lines from the file and save it, when it get's smaller. When I remove lines and save the file my index.html gets truncated,  the end of the file is missing. The word I added to the end is no longer in the file.

**Comment by kung-fu-tzu** [comment]
Related discussion at vagrant: https://github.com/mitchellh/vagrant/issues/2533
And a maybe working solution: https://github.com/mitchellh/vagrant/issues/2533#issuecomment-42606883

**Comment by adambiggs** [comment]
I came across this thread which also seems relevant: https://communities.vmware.com/thread/498273

**Comment by alexisvincent** [comment]
I remember there being an issue with virtual box and parallels not handling sendfile correctly. This lead to a host of issues. Could a similar thing be happening here?

**Comment by adambiggs** [comment]
I found a simpler way to reproduce the issue (no need to involve Docker), and wrote up some test cases below. Hope this helps to resolve this issue, since the VMware Fusion driver is nearly useless for local development until this is fixed. :cry:

#### Setup

Create a new VMware Fusion machine, SSH into it, and `cd` to your shared user directory: 
```
$ docker-machine create --driver vmwarefusion test
$ docker-machine ssh test
docker@test:~$ cd /Users/adam
```

In a different terminal, create a new test file anywhere in your user directory:
```
$ echo "initial content" > ~/test.txt
```

Check the file contents inside the VM (it should look good): 
```
docker@test:/Users/adam$ cat test.txt
initial content
```


#### Test 1: Modifying Contents

Modify the file on the host, *but use the same amount of characters*:
```
$ echo "updated content" > ~/test.txt
```

File contents don't change in the VM:
```
docker@test:/Users/adam$ cat test.txt
initial content
```

Even though the last-modified time matches:
```
$ ls -lT test.txt
-rw-r--r--  1 adam  staff  16 17 Jul 11:14:11 2015 test.txt
```
```
docker@test:/Users/adam$ ls -le test.txt
-rw-r--r--    1 501      20              16 Fri Jul 17 18:14:11 2015 test.txt
```

#### Test 2: Increase File Size

Modify the file, adding more characters:
```
$ echo "updated extra content" > test.txt
```

Wait a second... That worked?! :confused:
```
docker@test:/Users/adam$ cat test.txt
updated extra content
```

**Modifying a file seems to work if the file size is increased.** This has changed since I originally reported the issue... Possibly because I updated to v1.7.1 of boot2docker.iso?


#### Test 3: Decreasing File Size

Modify the file again, using less characters:
```
$ echo "shorter content" > test.txt
```

File length is updated to 16B in both places, but in the VM the file contains 16B of the old content:
```
docker@test:/Users/adam$ cat test.txt
updated extra co
```

**Comment by caleb** [comment]
I know it's not a docker-machine bug, but it is really unfortunate. I got all excited when I switched from boot2docker-cli to docker-machine because of the VMWare Fusion support, but then I ran into this :(

The VMWare thread linked here suggests that it is fixed VMWare Fusion, but I'm running into this problem on 7.1.2. Might it be because boot2docker.iso uses open vmware tools, and perhaps the bug is fixed in the non-open vmware tools?

Has anybody found a workaround that fixes this with docker-machine?

**Comment by caleb** [comment]
I tried a Debian 8.1 installation in VMWare Fusion with tools installed from the VMWare Fusion app (not open vmware tools), and I could not reproduce this file corruption bug.

So I used the generic Machine driver to install docker and wire everything up on my Mac, and everything seems to be working so far. Make sure you use an SSH key without a password until this is resolved: https://github.com/docker/machine/issues/1402

It's not as easy as the boot2docker setup, but it does seem to work.

**Comment by frapposelli** [comment]
All,

I'm working with the VMTools guys internally to resolve this, the fix is coming soon.

Some of the VMTools version shipped with our products are not affected (see @caleb post above), will let you know when it's fixed in `open-vm-tools`.

**Comment by steeve** [comment]
I did some testing, and I've observed the following behaviour:
- change is only detected if the file grows, even by 1 byte
- change is never detected when the file shrinks

My current workaround is to happen a empty line in the file, save it, remove it, and save it.

**Comment by steeve** [comment]
hey @frapposelli, any news on your front? thanks!

**Comment by adambiggs** [comment]
So far it looks like this is fixed in Machine v0.4.0-rc1!

**Comment by steeve** [comment]
I was under the impression this was due to openvm-tools not machine?

On Monday, July 27, 2015, Adam Biggs <notifications@github.com> wrote:

> So far it looks like this is fixed in Machine v0.4.0-rc1!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/machine/issues/1460#issuecomment-125297301>.
>


-- 
Steeve Morin
twitter.com/steeve
github.com/steeve
linkd.in/smorin

**Comment by frapposelli** [comment]
the new version of boot2docker (ng) is using the official debian packages for open-vm-tools (much older version that doesn't have that specific bug), I haven't tried it yet but v0.4.0-rc1 is most likely using the new b2d version.

**Comment by adambiggs** [comment]
Used v0.4.0-rc1 for a full work day yesterday with no problems, so I think it's safe to close this issue. Thanks all!

**Comment by adambiggs** [comment]
This is broken again in v0.4.0-rc2, because they reverted back to the old boot2docker version... Reopening this issue for visibility, but at least we know the fix should land in the next official version of Machine.

**Comment by caleb** [comment]
Doh! I upgraded my docker-machine and ran into this again. Thanks for the heads up. I'll revert back to v0.4.0-rc1 until it's sorted out.

Thanks!

**Comment by alexisvincent** [comment]
What is the current status on this? ETA?

**Comment by adambiggs** [comment]
Looks like we're all waiting for the first release of [docker/boot2docker](https://github.com/docker/boot2docker), but I'd love an ETA from somebody at Docker as well.

**Comment by frapposelli** [comment]
Quick update for all those who are waiting for a fix on this.

We have a new version of `open-vm-tools` with the bug fixed coming next week, once it's out, I will integrate with boot2docker (both old and new), as this new `open-vm-tools` version is using the new FUSE-based shared folders implementation is going to be much easier to move to new kernel versions without impacting the functionality.

Once a fixed version of boot2docker is ready for consumption I will let you guys know.

**Comment by alexisvincent** [comment]
:)

**Comment by donnex** [comment]
Sweet! In my short tests with Fusion I got greatly increased performance compared to Virtual Box.

Looking forward for a fix for this as well.

**Comment by vovimayhem** [comment]
+1

**Comment by yrashk** [comment]
Any updates on this? cc @frapposelli

**Comment by frapposelli** [comment]
@yrashk yes! `open-vm-tools` with the FUSE based shared folder implementation were released yesterday (https://github.com/vmware/open-vm-tools/tree/stable-10.0.x), I'll look into integrating them with Boot2Docker asap.

**Comment by bradgessler** [comment]
@frapposelli you are awesome. This is the last blocker that's preventing me from rolling out docker-machine to my dev team.

**Comment by andrerom** [comment]
:+1:

This issue has been around with vmware for years, making shared folder unusable for most dev use, just some:
- http://symfony2.narkive.com/VNOz4RHU/weird-error-on-cache-files
- https://communities.vmware.com/message/2326272
- https://github.com/mitchellh/vagrant/issues/2768
- http://stackoverflow.com/questions/6384088/weird-php-parse-syntax-errors-with-vmware-6-5-php-5-3-3-symfony-1-4
- http://share.ez.no/forums/ez-publish-5-platform/installation-saga
- http://www.stuartcarnie.com/2012/10/fixing-vmware-fusion-permissions-for.html

But even with this fixed I guess NFS is probably still the [best option](http://mitchellh.com/comparing-filesystem-performance-in-virtual-machines) for dev use.

**Comment by vovimayhem** [comment]
Cool :) Anyways @andrerom, is there any way to make docker (and docker-machine / docker-compose) mount volumes via NFS, without resorting to do it manually?

**Comment by andrerom** [comment]
I don't think so, not sure where that would be configured even with Docker Machine, which would be the part of the toolbox to deal with this.

**Comment by vovimayhem** [comment]
@andrerom: Have you checked out this? I've just stumbled upon this right now...
- [Docker Plugins](https://github.com/docker/docker/blob/master/docs/extend/plugins.md#finding-a-plugin)
- [Convoy (NFS example)](https://github.com/rancher/convoy#nfs)
- [An example using "rexray" and compose](https://github.com/kacole2/wordpress-compose/blob/master/docker-compose.yml)

**Comment by andrerom** [comment]
Not yet, but will have a look when this topic pops up, thanks :+1:

**Comment by caleb** [comment]
I recently discovered a tool like boot2docker that uses docker-machine called dinghy. It automatically runs a userland NFS server on your mac, and provides some other services like a friendly DNS server to use to access your images:

https://github.com/codekitchen/dinghy

It uses docker-machine to actually create and manage the linux machine. I've been using a generic docker-machine that I set up with debian and NFS, but I'm giving this a spin now.

**Comment by frapposelli** [comment]
Just a quick update, I submitted the relevant PRs to boot2docker/boot2docker#1054 and docker/machine #1816 to fix this issue.

This should remove the corruption problem entirely and give us an easy upgrade path to newer kernels as we've removed the dependency on a compiled kernel module.

**Comment by yrashk** [comment]
@frapposelli can't thank you enough!

**Comment by vovimayhem** [comment]
@frapposelli I've just pulled today both boot2docker & docker-machine from master, so I could do a test run... I created a docker machine using a boot2docker iso built from the boot2docker master.

What I saw is that whenever I mount a host directory into, and start adding files, those files can't be seen on the container... if I destroy the container, and start another one with the same mounted host directory, files start to appear, though not all the files.

Can you replicate this? Or is it just me?

**Comment by vovimayhem** [comment]
@frapposelli nevermind... I screwed with the docker-machine binary file, and where using the current version (0.4.1) instead of the one I built.

It is working like a charm.

**Comment by frapposelli** [comment]
@vovimayhem awesome, thanks for testing this!

@nathanleclaire both PRs were merged so we should be good with this.

**Comment by nathanleclaire** [comment]
@frapposelli Nice!

**Comment by nathanleclaire** [comment]
All, I'm closing this for now as it should be fixed in the next release, but please re-open the issue and/or open a new one if issues persist.

**Comment by nathanleclaire** [comment]
In the interim, maybe we can post a URL to an ISO with the fixes so that users can experiment with `--vmwarefusion-boot2docker-url`.

**Comment by nathanleclaire** [comment]
Anyone interested, I have compiled a `boot2docker.iso` from current master (`11ef978a79cb491d959e40ea390ed8045645615d`) and you can try it out like so:

```console
$ docker-machine create \
    -d vmwarefusion \
    --vmwarefusion-boot2docker-url https://s3-us-west-1.amazonaws.com/b2d.isos/boot2docker.iso \
    nocorrupt
```

**Comment by vovimayhem** [comment]
@nathanleclaire will not work with docker-machine 0.4.1. I had to compile docker-machine from master, following the [collaboration guide's instructions](https://github.com/docker/machine/blob/master/CONTRIBUTING.md)

I guess the fix will show up on the next release of docker-machine.

**Comment by nathanleclaire** [comment]
@vovimayhem Ah, gotcha.  I had forgotten that we merged some new changes to the Fusion driver upstream.

**Comment by nathanleclaire** [comment]
Can confirm that the mount seems to be working fine for me with current master.

**Comment by frapposelli** [comment]
Just as a note:

Releases from docker/machine@c3ed882b7612124ca8f050ff0275660082a41651 up work with boot2docker ISOs built from master (the one with FUSE shared folders) and previous releases as well (with the old kernel module).

It obviously doesn't work the other way around: boot2docker ISO from current master with `docker-machine` binaries built before docker/machine@c3ed882b7612124ca8f050ff0275660082a41651 will not work.

**Comment by seafoodbuffet** [comment]
Is there a date on when an official release is slated to include changes from this ticket?

**Comment by nathanleclaire** [comment]
@seafoodbuffet It will be in about two weeks, we are tied to the release cadence of the Docker Engine.

**Comment by donnex** [comment]
Any idea if this issue is resolved in the newly released Docker 1.9?

**Comment by seafoodbuffet** [comment]
I note that the latest docker-toolbox 1.9 release: https://github.com/docker/toolbox/releases
DOES include docker-machine 0.5.0 which has included this fix for some time. I'm betting that upgrading to the latest docker-toolbox should address this issue.

**Comment by adambiggs** [comment]
Sorry to say, but I think this issue might need to be reopened...

I've been trying to work with the latest docker-toolbox 1.9 release in VMware Fusion. But NPM is throwing `Unexpected end of input` errors when I try to run `$ npm install` in a host volume.

The same command works fine inside the container without a host volume mapping. But as soon as I use a host volume, NPM throws errors like this:

```
ci@8ef632634afa:~/build$ npm install
npm ERR! Linux 4.1.12-boot2docker
npm ERR! argv "node" "/usr/local/bin/npm" "install"
npm ERR! node v0.12.7
npm ERR! npm  v2.14.5
npm ERR! file /home/ci/build/node_modules/gulp-conventional-changelog/package.json
npm ERR! code EJSONPARSE

npm ERR! Failed to parse json
npm ERR! Unexpected end of input at 47:10
npm ERR!     "jshi
npm ERR!          ^
npm ERR! File: /home/ci/build/node_modules/gulp-conventional-changelog/package.json
npm ERR! Failed to parse package.json data.
npm ERR! package.json must be actual JSON, not just JavaScript.
npm ERR!
npm ERR! This is not a bug in npm.
npm ERR! Tell the package author to fix their package.json file. JSON.parse

npm ERR! Please include the following file with any support request:
npm ERR!     /home/ci/build/npm-debug.log
```

**Comment by nathanleclaire** [comment]
cc @frapposelli

**Comment by frapposelli** [comment]
@adambiggs can you give me a `mount` and `df -h` of the VM? the corruption issue should be fixed in 0.5.0

**Comment by adambiggs** [comment]
```
                        ##         .
                  ## ## ##        ==
               ## ## ## ## ##    ===
           /"""""""""""""""""\___/ ===
      ~~~ {~~ ~~~~ ~~~ ~~~~ ~~~ ~ /  ===- ~~~
           \______ o           __/
             \    \         __/
              \____\_______/
 _                 _   ____     _            _
| |__   ___   ___ | |_|___ \ __| | ___   ___| | _____ _ __
| '_ \ / _ \ / _ \| __| __) / _` |/ _ \ / __| |/ / _ \ '__|
| |_) | (_) | (_) | |_ / __/ (_| | (_) | (__|   <  __/ |
|_.__/ \___/ \___/ \__|_____\__,_|\___/ \___|_|\_\___|_|
Boot2Docker version 1.9.0, build master : 16e4a2a - Tue Nov  3 19:49:22 UTC 2015
Docker version 1.9.0, build 76d6bc9
docker@lmpm-vmware:~$ mount
tmpfs on / type tmpfs (rw,relatime,size=3640972k)
proc on /proc type proc (rw,relatime)
sysfs on /sys type sysfs (rw,relatime)
devpts on /dev/pts type devpts (rw,relatime,mode=600,ptmxmode=000)
tmpfs on /dev/shm type tmpfs (rw,relatime)
fusectl on /sys/fs/fuse/connections type fusectl (rw,relatime)
/dev/sda1 on /mnt/sda1 type ext4 (rw,relatime,data=ordered)
cgroup on /sys/fs/cgroup type tmpfs (rw,relatime,mode=755)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,relatime,cpuset)
cgroup on /sys/fs/cgroup/cpu type cgroup (rw,relatime,cpu)
cgroup on /sys/fs/cgroup/cpuacct type cgroup (rw,relatime,cpuacct)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,relatime,blkio)
cgroup on /sys/fs/cgroup/memory type cgroup (rw,relatime,memory)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,relatime,devices)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,relatime,freezer)
cgroup on /sys/fs/cgroup/net_cls type cgroup (rw,relatime,net_cls)
cgroup on /sys/fs/cgroup/perf_event type cgroup (rw,relatime,perf_event)
cgroup on /sys/fs/cgroup/net_prio type cgroup (rw,relatime,net_prio)
cgroup on /sys/fs/cgroup/hugetlb type cgroup (rw,relatime,hugetlb)
vmhgfs-fuse on /mnt/hgfs type fuse.vmhgfs-fuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other)
vmhgfs-fuse on /Users type fuse.vmhgfs-fuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other)
/dev/sda1 on /mnt/sda1/var/lib/docker/aufs type ext4 (rw,relatime,data=ordered)
nsfs on /var/run/docker/netns/613ef796b6d4 type nsfs (rw)
docker@lmpm-vmware:~$ df -h
Filesystem                Size      Used Available Use% Mounted on
tmpfs                     3.5G    123.8M      3.4G   3% /
tmpfs                     1.9G         0      1.9G   0% /dev/shm
/dev/sda1                18.2G      3.6G     13.6G  21% /mnt/sda1
cgroup                    1.9G         0      1.9G   0% /sys/fs/cgroup
vmhgfs-fuse             809.5G    669.9G    139.6G  83% /mnt/hgfs
vmhgfs-fuse             809.5G    669.9G    139.6G  83% /Users
/dev/sda1                18.2G      3.6G     13.6G  21% /mnt/sda1/var/lib/docker/aufs
```

**Comment by frapposelli** [comment]
@adambiggs I cannot reproduce the corruption problem anymore, can you please provide me some repro steps?

**Comment by adambiggs** [comment]
@frapposelli sorry for the slow reply, I had a chance to test this again and have come up with steps to reproduce. I noticed this when trying to install NPM packages, but there may very well be other ways to reproduce it.

1. Create a bare-bones Node.js image: `$ echo "FROM node" > Dockerfile && docker build -t "vmware-test" .`
2. Start a container with a host volume: `$ docker run -ti --rm -v ${PWD}:/host-volume vmware-test bash`
3. Try installing a node package globally (it should work): `# npm install -g lodash`
4. Now try installing the same package locally in the host volume: `# cd /host-volume && npm install lodash`

When installing the package locally you should see the `EJSONPARSE Failed to parse json` error.

**Comment by adambiggs** [comment]
Interestingly, the error message is slightly different using the latest Node.js v5 image. This part in particular:

```
EPERM: operation not permitted, chown '/host-volume/node_modules/.staging/lodash-f4a0ee8d651a0e87283fd2e368346bf9/package.json'
```

**Comment by Korri** [comment]
I have the same exact issue on Virtuabox Windows 10, worked around it by disabling sendfile, but still pretty weird and anoying.

**Comment by donnex** [comment]
I just tried the new vmwarefusion version. I didn't notice any of the previous errors. The original problem seems to be gone.

What I did notice was that now vmwarefusion is very slow to detect file changes compared to virtualbox. I compared two different projects with mounted folders from the host system. Both of them ran watches on frontend files (css, js etc). When using vmwarefusion the file changes and updates where detected 5-10 sec slowe than on virtualbox. The difference was really noticeable.

**Comment by adambiggs** [comment]
For anybody who's still having issues, switching to NFS file sharing has solved all mo vmware issues.

There's a PR [over here](https://github.com/adlogix/docker-machine-nfs/pull/26) that got me up and running.

**Comment by ain** [comment]
I'm calling to reopen this ticket. I'm running `1.10.0` with respective boot2docker.iso and I can still reproduce the issue. File rights differ.

**Comment by frapposelli** [comment]
@ain please file another issue with reproduction steps.

**Comment by gravis** [comment]
I'm having the exact same behaviour with docker-machine `0.6.0`, docker `1.10.2`

**Comment by donnex** [comment]
Maybe this is a non issue with the new Docker for Mac? I haven't tried the beta yet but it sounds like it got better file sharing.

**Comment by ain** [comment]
@frapposelli I have described it in #1789. It's a very simple use-case and a blocker, unfortunately. 

I have just reproduced it on Docker Machine `0.7.0` and Docker Engine `1.11.0`.

**Comment by karlkfi** [comment]
I'm actually seeing these exact same corruption/file-length symptoms on Mac with VirtualBox `5.0.20`, Docker Machine `0.7.0`, and Docker Engine `1.11.1`.

I get the feeling it's not VMware Fusion specific.

**Comment by wildsurfer** [comment]
Exactly the same issue. VirtualBox 5.0.24, docker-machine 0.7.0, Mac
